File: backend/src/services/redis.ts
import { Redis } from 'ioredis';
import { createAdapter } from '@socket.io/redis-adapter';
import type { Server as SocketIOServer } from 'socket.io';

const REDIS_URL = process.env.REDIS_URL!;
if (!REDIS_URL) {
  console.error('[Fatal] REDIS_URL must be set');
  process.exit(1);
}

// ─── Core Redis Client ───────────────────────────────────────────
export const redis = new Redis(REDIS_URL, {
  retryStrategy: (times) => Math.min(times * 50, 2000),
  maxRetriesPerRequest: 3
});

redis.on('error', (err) => {
  console.error('[Redis] Connection error:', err.message);
});

redis.on('connect', () => {
  console.log('[Redis] Connected successfully');
});

// ─── Cache Helpers ───────────────────────────────────────────────
export const cache = {
  async get<T>(key: string): Promise<T | null> {
    const data = await redis.get(key);
    if (!data) return null;
    try {
      return JSON.parse(data) as T;
    } catch {
      return null;
    }
  },

  async set(key: string, value: unknown, ttlSeconds = 300): Promise<void> {
    await redis.setex(key, ttlSeconds, JSON.stringify(value));
  },

  async del(key: string): Promise<void> {
    await redis.del(key);
  },

  async keys(pattern: string): Promise<string[]> {
    return redis.keys(pattern);
  }
};

// ─── Presence & Socket Registry (Redis-backed for multi-node scaling) ──
const SOCKET_TTL = 3600; // 1 hour
const PRESENCE_TTL = 30; // 30 seconds (strict heartbeat)
const HEARTBEAT_INTERVAL = 10000; // 10 seconds

export type PresenceStatus = 'online' | 'away' | 'typing' | 'offline';

export interface PresenceData {
  status: PresenceStatus;
  lastSeen: string;
  displayName: string;
}

export const presence = {
  // Store a single active socket per user (latest wins for direct messaging)
  async setSocket(username: string, socketId: string): Promise<void> {
    await redis.setex(`rina:socket:${username}`, SOCKET_TTL, socketId);
  },

  async getSocket(username: string): Promise<string | null> {
    return redis.get(`rina:socket:${username}`);
  },

  async delSocket(username: string): Promise<void> {
    await redis.del(`rina:socket:${username}`);
  },

  // Multi-tab tracking: add socket to user's set
  async addUserSocket(username: string, socketId: string): Promise<void> {
    await redis.sadd(`rina:sockets:${username}`, socketId);
    await redis.expire(`rina:sockets:${username}`, SOCKET_TTL);
  },

  async removeUserSocket(username: string, socketId: string): Promise<number> {
    await redis.srem(`rina:sockets:${username}`, socketId);
    const remaining = await redis.scard(`rina:sockets:${username}`);
    if (remaining === 0) {
      await redis.del(`rina:sockets:${username}`);
    }
    return remaining;
  },

  async getUserSocketCount(username: string): Promise<number> {
    return redis.scard(`rina:sockets:${username}`);
  },

  async setStatus(username: string, data: PresenceData): Promise<void> {
    await redis.setex(`rina:presence:${username}`, PRESENCE_TTL, JSON.stringify(data));
  },

  async getStatus(username: string): Promise<PresenceData | null> {
    const data = await redis.get(`rina:presence:${username}`);
    if (!data) return null;
    try {
      return JSON.parse(data) as PresenceData;
    } catch {
      return null;
    }
  },

  async setHeartbeat(username: string): Promise<void> {
    await redis.setex(`rina:heartbeat:${username}`, PRESENCE_TTL, Date.now().toString());
  },

  async getHeartbeat(username: string): Promise<number | null> {
    const ts = await redis.get(`rina:heartbeat:${username}`);
    return ts ? parseInt(ts, 10) : null;
  }
};

// ─── Socket.io Redis Adapter ─────────────────────────────────────
export function setupSocketAdapter(io: SocketIOServer): void {
  const pubClient = new Redis(REDIS_URL, {
    retryStrategy: (times) => Math.min(times * 50, 2000),
    maxRetriesPerRequest: 3
  });
  const subClient = pubClient.duplicate();

  pubClient.on('error', (err) => {
    console.error('[Redis Pub] Connection error:', err.message);
  });

  subClient.on('error', (err) => {
    console.error('[Redis Sub] Connection error:', err.message);
  });

  io.adapter(createAdapter(pubClient, subClient));
  console.log('[Socket.io] Redis adapter initialized');
}

export { HEARTBEAT_INTERVAL };

File: go-backend/internal/services/redis.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
	"rina-backend/internal/config"
)

var rdb *redis.Client

func ConnectRedis() error {
	cfg := config.Get()
	opt, err := redis.ParseURL(cfg.RedisURL)
	if err != nil {
		return fmt.Errorf("parse redis url: %w", err)
	}
	rdb = redis.NewClient(opt)
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	if err := rdb.Ping(ctx).Err(); err != nil {
		return fmt.Errorf("ping redis: %w", err)
	}
	return nil
}

func GetRedis() *redis.Client {
	return rdb
}

func CloseRedis() {
	if rdb != nil {
		rdb.Close()
	}
}

// ─── Cache Helpers ─────────────────────────────────────────────────

func CacheGet[T any](ctx context.Context, key string) (T, error) {
	var result T
	data, err := rdb.Get(ctx, key).Result()
	if err != nil {
		return result, err
	}
	if err := json.Unmarshal([]byte(data), &result); err != nil {
		return result, err
	}
	return result, nil
}

func CacheSet(ctx context.Context, key string, value interface{}, ttl time.Duration) error {
	data, err := json.Marshal(value)
	if err != nil {
		return err
	}
	return rdb.Set(ctx, key, data, ttl).Err()
}

func CacheDel(ctx context.Context, key string) error {
	return rdb.Del(ctx, key).Err()
}

// ─── Presence ──────────────────────────────────────────────────────

const (
	SocketTTL       = 3600 * time.Second
	PresenceTTL     = 30 * time.Second
	HeartbeatTTL    = 10 * time.Second
)

type PresenceData struct {
	Status      string `json:"status"`
	LastSeen    string `json:"lastSeen"`
	DisplayName string `json:"displayName"`
}

func SetSocket(ctx context.Context, username, socketID string) error {
	return rdb.Set(ctx, fmt.Sprintf("rina:socket:%s", username), socketID, SocketTTL).Err()
}

func GetSocket(ctx context.Context, username string) (string, error) {
	return rdb.Get(ctx, fmt.Sprintf("rina:socket:%s", username)).Result()
}

func DelSocket(ctx context.Context, username string) error {
	return rdb.Del(ctx, fmt.Sprintf("rina:socket:%s", username)).Err()
}

func AddUserSocket(ctx context.Context, username, socketID string) error {
	key := fmt.Sprintf("rina:sockets:%s", username)
	if err := rdb.SAdd(ctx, key, socketID).Err(); err != nil {
		return err
	}
	return rdb.Expire(ctx, key, SocketTTL).Err()
}

func RemoveUserSocket(ctx context.Context, username, socketID string) error {
	return rdb.SRem(ctx, fmt.Sprintf("rina:sockets:%s", username), socketID).Err()
}

func SetStatus(ctx context.Context, username string, data PresenceData) error {
	b, _ := json.Marshal(data)
	return rdb.Set(ctx, fmt.Sprintf("rina:status:%s", username), b, PresenceTTL).Err()
}

func GetStatus(ctx context.Context, username string) (*PresenceData, error) {
	data, err := rdb.Get(ctx, fmt.Sprintf("rina:status:%s", username)).Result()
	if err != nil {
		return nil, err
	}
	var pd PresenceData
	if err := json.Unmarshal([]byte(data), &pd); err != nil {
		return nil, err
	}
	return &pd, nil
}

func SetHeartbeat(ctx context.Context, username string) error {
	return rdb.Set(ctx, fmt.Sprintf("rina:heartbeat:%s", username), time.Now().Format(time.RFC3339), HeartbeatTTL).Err()
}

func GetUserSockets(ctx context.Context, username string) ([]string, error) {
	return rdb.SMembers(ctx, fmt.Sprintf("rina:sockets:%s", username)).Result()
}
